### PR TITLE
Implement PeriodIndex.difference without object-dtype cast

### PR DESCRIPTION
**Pull request by jbrockmendel**
Analogous to #30666.  After this we'll be able to share some code between the set operations.

The edit in pd._testing is mostly unrelated.  Both are motivated by tracking down the places where object-dtype ndarray is passed to PeriodIndex._shallow_copy.

### CODE REVIEW

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -594,13 +595,13 @@ def get_indexer_non_unique(self, target):
...
-        wrap Index._get_unique_index to handle NaT
-        """
-        res = super()._get_unique_index(dropna=dropna)
-        if dropna:
-            res = res.dropna()
-        return res
+        if self.is_unique and not dropna:
+            return self
+
+        result = self._data.unique()
```
Comment (jreback) [inline pandas/core/indexes/period.py]: why is this not self.unique()?

File: pandas/core/indexes/period.py [hunk 2]
```diff
@@ -818,6 +819,29 @@ def intersection(self, other, sort=False):
         result = self._shallow_copy(np.asarray(i8result, dtype=np.int64), name=res_name)
         return result
 
+    def difference(self, other, sort=None):
```
Comment (jreback) [inline pandas/core/indexes/period.py]: eventtually going to share this?

File: pandas/_testing.py [hunk 3]
```diff
@@ -619,7 +619,10 @@ def _get_ilevel_values(index, level):
         # accept level number only
         unique = index.levels[level]
         level_codes = index.codes[level]
-        filled = take_1d(unique.values, level_codes, fill_value=unique._na_value)
+        if is_extension_array_dtype(unique):
```
Comment (jreback) [inline pandas/_testing.py]: this is not as nice (though its only for testing), won't unique.take(...) work here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/period.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: yes, all the PI set operations will end up sharing code

File: pandas/_testing.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_testing.py]: ill give it a shot

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: Because we want to pass an EA to shallow_copy

Comment (jreback) [reply pandas/core/indexes/period.py]: question here

File: pandas/_testing.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_testing.py]: Turns out this unique.take(...) wont work here.  the trouble is that we are calling _shallow_copy below (presumably for perf) and using unique.take ends up passing the wrong types to shallow_copy

File: pandas/core/indexes/period.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: calling self.unique returns an Index, while we want a PeriodArray to pass to _shallow_copy below

File: pandas/_testing.py (hunk 3 above)
Comment (jreback) [reply pandas/_testing.py]: > using unique.take ends up passing the wrong types to shallow_copy

then i'd rather not change this, take_1d just does the right thing, so why change it?

Comment (jbrockmendel) [reply pandas/_testing.py]: Looks like passing unique._values instead of unique.values fixes the problem cleanly

Comment (jreback) [reply pandas/_testing.py]: ok great

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on green.

**Comment by jreback** [comment]
thanks

**Pull request merged**